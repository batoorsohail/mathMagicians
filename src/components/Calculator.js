import React, { useState } from 'react';
import calculate from '../logic/calculate';

function Calculator() {
  const [state, setState] = useState({ total: 0, next: null, operation: null });

  const clickHandler = (e) => {
    setState((state) => calculate(state, e.target.value));
  };
  return (
    <section className="calculator-container">
      <div className="result">
        <p>
          {state.total}
          {state.operation}
          {state.next}
        </p>
      </div>
      <div className="calculator">
        <button type="button" className="button" onClick={clickHandler} value="AC">AC</button>
        <button type="button" className="button" onClick={clickHandler} value="+/-">+/-</button>
        <button type="button" className="button" onClick={clickHandler} value="%">%</button>
        <button type="button" className="operator" onClick={clickHandler} value="รท">รท</button>
        <button type="button" className="button" onClick={clickHandler} value="7">7</button>
        <button type="button" className="button" onClick={clickHandler} value="8">8</button>
        <button type="button" className="button" onClick={clickHandler} value="9">9</button>
        <button type="button" className="operator" onClick={clickHandler} value="x">x</button>
        <button type="button" className="button" onClick={clickHandler} value="4">4</button>
        <button type="button" className="button" onClick={clickHandler} value="5">5</button>
        <button type="button" className="button" onClick={clickHandler} value="6">6</button>
        <button type="button" className="operator" onClick={clickHandler} value="-">-</button>
        <button type="button" className="button" onClick={clickHandler} value="3">1</button>
        <button type="button" className="button" onClick={clickHandler} value="2">2</button>
        <button type="button" className="button" onClick={clickHandler} value="1">3</button>
        <button type="button" className="operator" onClick={clickHandler} value="+">+</button>
        <button type="button" className="button btn-0" onClick={clickHandler} value="0">0</button>
        <button type="button" className="button" onClick={clickHandler} value=".">.</button>
        <button type="button" className="operator" onClick={clickHandler} value="=">=</button>
      </div>
    </section>
  );
}

export default Calculator;
