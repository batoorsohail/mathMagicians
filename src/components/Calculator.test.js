import { render, screen, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import Calculator from "./Calculator";
it("renders correctly", () => {
  const tree = renderer.create(<Calculator />).toJSON();
  expect(tree).toMatchSnapshot();
});
describe("User interactions", () => {
  test("if the correct special key is passed for clearing data", () => {
    render(<Calculator />);
    let buttonElement;
    act(() => {
      buttonElement = screen.getByText("AC");
      userEvent.click(buttonElement);
    });
    const outputElement = buttonElement;
    expect(outputElement.textContent).toBe("AC");
  });
  test("if the right special key is passed for equality", () => {
    render(<Calculator />);
    let buttonElement;
    act(() => {
      buttonElement = screen.getByText("=");
      userEvent.click(buttonElement);
    });
    const outputElement = buttonElement;
    expect(outputElement.textContent).toBe("=");
  });
  test("if the right special key is passed for sign change", () => {
    render(<Calculator />);
    let buttonElement;
    act(() => {
      buttonElement = screen.getByText("+/-");
      userEvent.click(buttonElement);
    });
    const outputElement = buttonElement;
    expect(outputElement.textContent).toBe("+/-");
  });
  test("if the right operation is passed for addition", () => {
    render(<Calculator />);
    let buttonElement;
    act(() => {
      buttonElement = screen.getByText("+");
      userEvent.click(buttonElement);
    });
    const outputElement = buttonElement;
    expect(outputElement.textContent).toBe("+");
  });
  test("if the right operation is passed for subtraction", () => {
    render(<Calculator />);
    let buttonElement;
    act(() => {
      buttonElement = screen.getByText("-");
      userEvent.click(buttonElement);
    });
    const outputElement = buttonElement;
    expect(outputElement.textContent).toBe("-");
  });
  test("if the right operation is passed for multiplication", () => {
    render(<Calculator />);
    let buttonElement;
    act(() => {
      buttonElement = screen.getByText("x");
      userEvent.click(buttonElement);
    });
    const outputElement = buttonElement;
    expect(outputElement.textContent).toBe("x");
  });
  test("if the right operation is passed for division", () => {
    render(<Calculator />);
    let buttonElement;
    act(() => {
      buttonElement = screen.getByText("รท");
      userEvent.click(buttonElement);
    });
    const outputElement = buttonElement;
    expect(outputElement.textContent).toBe("รท");
  });
  test("if the right operation is passed for modulo", () => {
    render(<Calculator />);
    let buttonElement;
    act(() => {
      buttonElement = screen.getByText("%");
      userEvent.click(buttonElement);
    });
    const outputElement = buttonElement;
    expect(outputElement.textContent).toBe("%");
  });
});
