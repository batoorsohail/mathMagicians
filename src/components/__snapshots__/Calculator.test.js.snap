// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="calculator-container"
>
  <div
    className="result"
  >
    <p>
      0
    </p>
  </div>
  <div
    className="calculator"
  >
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="AC"
    >
      AC
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="+/-"
    >
      +/-
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="%"
    >
      %
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="รท"
    >
      รท
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="7"
    >
      7
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="8"
    >
      8
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="9"
    >
      9
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="x"
    >
      x
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="4"
    >
      4
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="5"
    >
      5
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="6"
    >
      6
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="-"
    >
      -
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="3"
    >
      1
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="2"
    >
      2
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="1"
    >
      3
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="+"
    >
      +
    </button>
    <button
      className="button btn-0"
      onClick={[Function]}
      type="button"
      value="0"
    >
      0
    </button>
    <button
      className="button"
      onClick={[Function]}
      type="button"
      value="."
    >
      .
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="="
    >
      =
    </button>
  </div>
</section>
`;
